### PR TITLE
UCT/API: uct_iface_is_reachable() interface changed.

### PR DESCRIPTION
**Pull request by shssf**
- added two parameters to the function uct_iface_is_reachable
- parameter names are discussable (tl_iface is better)
- Perhaps, it is better to change "const type *var" to "const type *
  const var" to guarantee no pointer changes
- active_messages example remains with old behavior to keep the code simple

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2142/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1492/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2143/ for details.

**Comment by yosefe** [comment]
@shamisp pls review

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2147/ for details.

**Comment by yosefe** [comment]
:+1:

**Comment by shssf** [comment]
Just resolved small branch conflicts with newly merged changes

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2157/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1502/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2160/ for details.

**Comment by shamisp** [comment]
do we really need remote ep ?

**Comment by shamisp** [comment]
Also. Please add justification for this API change. What kinda problem does it solve and why it can not be solved with previous API.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2161/ for details.

**Comment by shssf** [comment]
@shamisp Currently `dev_addr` used only in interface reachability procedure. In `SELF` interface no device exists and `iface_addr` used only. To allow flexible reachability procedure we need all these addresses because only specific transport could know all its requirements.
For example, `SELF` interface don’t need `dev_addr` but required `iface_addr` to complete `uct_iface_is_reachable()` call.
Also, we have `ep_addr` that currently doesn’t need to the procedure but might be required later. I added `ep_addr` to the function call to extend the interface once.

**Comment by shssf** [comment]
Related issue [878](https://github.com/openucx/ucx/issues/878)

**Comment by shamisp** [comment]
@shssf - does it mean that you can pass NULL for some of the parameters ? If so, it has to be explicitly specified in the API documentation

**Comment by shssf** [comment]
@shamisp Yes, I used NULL as parameters in active_messages.c example in this PR.
Thank you, I will add documentation about this.

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -693,11 +693,14 @@ ucs_status_t uct_iface_get_address(uct_iface_h iface, uct_iface_addr_t *addr);
  * it does not detect issues such as network mis-configuration or lack of connectivity.
  *
  * @param [in]  iface      Interface to check reachability from.
- * @param [in]  addr       Address to check reachability to.
+ * @param [in]  dev_addr   Device address to check reachability to.
+ * @param [in]  ep_addr    Remote endpoint address to check reachability to.
+ * @param [in]  iface_addr Interface address to check reachability to.
```
Comment (shamisp) [inline src/uct/api/uct.h]: The documentation has to be updated with the description of NULL parameter.

**Comment by yosefe** [comment]
we need to define whether we have to pass all addresses, or are allowed to pass NULL

**Comment by shssf** [comment]
@shamisp I added a small \@note in documentation about ability to pass NULL as parameter.

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -693,11 +693,17 @@ ucs_status_t uct_iface_get_address(uct_iface_h iface, uct_iface_addr_t *addr);
...
+ *
+ * @note One or more addresses can be NULL depends on interface specific.
+ *       If all addresses are NULL the return value will be 0.
```
Comment (shamisp) [inline src/uct/api/uct.h]: We had this discussion on the call. It is not clear when we have to pass NULL. I think @yosefe suggested to remove EP because it is really confusing

Comment (shssf) [inline src/uct/api/uct.h]: Sorry, I missed the design call. 
Ok, I will remove `ep_addr`.
I don't understand, what does it mean "we have to pass NULL". We don't have to pass NULL to the function, but we can if we know what we are doing. It depends on interface specific requirements. For example, if the interface supports UCT_IFACE_FLAG_CONNECT_TO_IFACE only we don't need `dev_addr` (SELF specific).

Comment (shamisp) [inline src/uct/api/uct.h]: Before you remove, talk with yossi.

Comment (shamisp) [inline src/uct/api/uct.h]: Well, for certain interface the EP is really optional. You can check connectivity to EP or to iface. It is not clear when you have to pass what ? What if I pass NULL and the remote EP is not null ? What does it check and return ?

Comment (shssf) [inline src/uct/api/uct.h]: Thank you. I will talk to Yossi about this.
In case of the question "What if I pass NULL and the remote EP is not null" I would say that the caller algorithm don't want to connect to `EP` but want to connect to `iface`. As I understand, this is the main idea in having these two addresses (`dev_add` and `iface_addr`) - to provide better flexibility in connection procedure. In one hand, we have algorithm which wants to select interface/ep to work and, in other hand, we have the interface which should decide workability with the algorithm. If the interface gets NULL in the pointer it required - the answer `false` (I don't want to be selected).
Sorry, perhaps I'm wrong and I need to learn an other interfaces logic more.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1514/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2170/ for details.

File: src/uct/api/uct.h (hunk 2 above)
Comment (yosefe) [inline src/uct/api/uct.h]: yep, lets remove the `ep_addr` and keep only `dev_addr` and `iface_addr`:
* `dev_addr` may be NULL if `iface_attr.dev_addr_len == 0`, and has to be non-NULL otherwise.
* `iface_addr` may be NULL if `iface_attr.iface_addr_len == 0`, and has to be non-NULL otherwise.

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -116,7 +116,7 @@ static int ucp_wireup_is_reachable(ucp_worker_h worker, ucp_rsc_index_t rsc_inde
 {
     ucp_context_h context = worker->context;
     return (context->tl_rscs[rsc_index].tl_name_csum == ae->tl_name_csum) &&
-           uct_iface_is_reachable(worker->ifaces[rsc_index], ae->dev_addr);
+           uct_iface_is_reachable(worker->ifaces[rsc_index], ae->dev_addr, ae->iface_addr);
```
Comment (shssf) [inline src/ucp/wireup/select.c]: In this place I need to change algo to provide NULL according to the new interface requirements. It require loop-back interface changes. It will be done in separate PR.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1517/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2173/ for details.

File: src/ucp/wireup/select.c (hunk 3 above)
Comment (yosefe) [inline src/ucp/wireup/select.c]: we can say that "iface_addr *may* be NULL"
so we can just keep it this way:
`uct_iface_is_reachable(worker->ifaces[rsc_index], ae->dev_addr, ae->iface_addr);`

Comment (shssf) [inline src/ucp/wireup/select.c]: @yosefe `dev_addr` and `iface_addr` are pointers to the addresses structure. They always points to the memory as I described in [878](https://github.com/openucx/ucx/issues/878) and contains trash if add_len == 0. I think we have to set them to NULL explicitly here or in address_unpuck procedure. The second one requires memory allocation for the addresses.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2176/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1521/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2178/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2179/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2182/ for details.

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -693,11 +693,16 @@ ucs_status_t uct_iface_get_address(uct_iface_h iface, uct_iface_addr_t *addr);
  * it does not detect issues such as network mis-configuration or lack of connectivity.
  *
  * @param [in]  iface      Interface to check reachability from.
- * @param [in]  addr       Address to check reachability to.
+ * @param [in]  dev_addr   Device address to check reachability to. It may be NULL
+ *                         if iface_attr.dev_addr_len == 0, and has to be non-NULL otherwise.
+ * @param [in]  iface_addr Interface address to check reachability to. It may
```
Comment (shamisp) [inline src/uct/api/uct.h]: I think it "must be" and not "may be". If the iface_attr.iface_addr_len == 0 and user pass some value it most likely to lead to an error, right ?

Comment (shssf) [inline src/uct/api/uct.h]: @shamisp Yes, you are right. Would you like me to change the documentation in the way like:
`Device address to check reachability to. It is NULL if iface_attr.dev_addr_len == 0, and must to be non-NULL otherwise.`

Comment (shamisp) [inline src/uct/api/uct.h]: looks good to me ! Please go ahead and update the documentation. thanks !

**Comment by shssf** [comment]
I expect it is ready to merge

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1528/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2187/ for details.
